File: A1/python/main.py
import sys

def dfs(level, total, ans):
    if total == 0:
        print( "(" +  ans + (")"*level) + ")" )
    elif level == 0:
        dfs(level+1, total-1, ans + "(")
    else:
        dfs(level+1, total-1, ans + "(")
        dfs(level-1, total, ans + ")")

def trees(n):
    dfs(0, n-1, "")

def main():
    print ("How many points do you have ?")
    x = int(sys.stdin.readline());
    trees(x)

if __name__ == "__main__":
    main()

File: A1/cpp/main.cpp
#include<iostream>
#include<string>
using namespace std;

void dfs(int level, int n, string ans)
{
    if(n == 0){
        cout << '(' << (ans + string(level, ')')) << ')' << endl;
    }
    else if(level == 0){
        dfs(level+1, n-1, ans + string(1, '('));
    }else{
        dfs(level+1, n-1, ans + string(1, '(') );
        dfs(level-1, n, ans + string(1, ')') );
    }
}

void trees(int n)
{
    dfs(0, n-1, "");
}

int main(int argc, char** argv)
{
    int n;
    cout<< "\nHow many points do you have?" << endl;
    cin >> n;
    trees(n);
    return 0;
}

